lib/usd/hdMaya/utils.cpp

```diff
@@ -0,0 +1,177 @@
+#include <pxr/usdImaging/usdImaging/textureUtils.h>
+#endif // USD_VERSION_NUM >= 1901
+
+#include <pxr/imaging/hdSt/textureResource.h>
+
+#include <maya/MPlugArray.h>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+namespace {
+
+#if USD_VERSION_NUM >= 1901
+class UdimTextureFactory : public GlfTextureFactoryBase {
+public:
+    virtual GlfTextureRefPtr New(
+        TfToken const& texturePath,
+        GlfImage::ImageOriginLocation originLocation =
+            GlfImage::OriginLowerLeft) const override {
+        const GlfContextCaps& caps = GlfContextCaps::GetInstance();
+        return GlfUdimTexture::New(
+            texturePath, originLocation,
+            UsdImaging_GetUdimTiles(texturePath, caps.maxArrayTextureLayers));
+    }
+
+    virtual GlfTextureRefPtr New(
+        TfTokenVector const& texturePaths,
+        GlfImage::ImageOriginLocation originLocation =
+            GlfImage::OriginLowerLeft) const override {
+        return nullptr;
+    }
+};
+#endif // USD_VERSION_NUM >= 1901
+
+} // namespace
+
+MObject GetConnectedFileNode(const MObject& obj, const TfToken& paramName) {
+    MStatus status;
+    MFnDependencyNode node(obj, &status);
+    if (ARCH_UNLIKELY(!status)) { return MObject::kNullObj; }
+    return GetConnectedFileNode(node, paramName);
+}
+
+MObject GetConnectedFileNode(
+    const MFnDependencyNode& node, const TfToken& paramName) {
+    MPlugArray conns;
+    node.findPlug(paramName.GetText(), true).connectedTo(conns, true, false);
+    if (conns.length() == 0) { return MObject::kNullObj; }
+    const auto ret = conns[0].node();
+    if (ret.apiType() == MFn::kFileTexture) { return ret; }
+    return MObject::kNullObj;
+}
+
+TfToken GetFileTexturePath(const MFnDependencyNode& fileNode) {
+    if (fileNode.findPlug(MayaAttrs::file::uvTilingMode, true).asShort() != 0) {
+        const TfToken ret{
+            fileNode.findPlug(MayaAttrs::file::fileTextureNamePattern, true)
+                .asString()
+                .asChar()};
+        return ret.IsEmpty()
+                   ? TfToken{fileNode
+                                 .findPlug(
+                                     MayaAttrs::file::
+                                         computedFileTextureNamePattern,
+                                     true)
+                                 .asString()
+                                 .asChar()}
+                   : ret;
+    } else {
+        const TfToken ret{
+            MRenderUtil::exactFileTextureName(fileNode.object()).asChar()};
+        return ret.IsEmpty()
+                   ? TfToken{fileNode
+                                 .findPlug(
+                                     MayaAttrs::file::fileTextureName, true)
+                                 .asString()
+                                 .asChar()}
+                   : ret;
+    }
+}
+
+std::tuple<HdWrap, HdWrap> GetFileTextureWrappingParams(
+    const MObject& fileObj) {
+    constexpr std::tuple<HdWrap, HdWrap> def{HdWrapClamp, HdWrapClamp};
```

Comment:
This line fails to compile on macOS (not in strict mode), details

/git/usd/maya-usd/lib/usd/hdMaya/utils.cpp:110:42: error: constexpr variable 'def' must be initialized by a constant expression
    constexpr std::tuple<HdWrap, HdWrap> def{HdWrapClamp, HdWrapClamp};
                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/git/usd/maya-usd/lib/usd/hdMaya/utils.cpp:110:42: note: non-constexpr constructor 'tuple<pxrInternal_v0_20__pxrReserved__::HdWrap, pxrInternal_v0_20__pxrReserved__::HdWrap, false, false>' cannot be used in a constant expression
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/tuple:723:9: note: declared here
        tuple(_Up&&... __u)

Reply:
Hmm, guess this is a difference between how clang and gcc handle constexpr? I'll take a look...

Reply:
So, apparently, tuples are strictly only constexpr for c++14, not c++11.  Apparently gcc still allows / makes them constexpr, though.

Regardless, compiling in optimized mode should treat const and constexpr in this situation identically:

https://godbolt.org/z/C6wx5U

...so I'll just change to const.

Reply:
Fixed!